src/test/java/org/jenkinsci/plugins/workflow/graph/FlowNodeTest.java

```diff
@@ -121,4 +127,303 @@ private static void assertActiveSteps(WorkflowRun b, String... expected) {
         assertEquals(Arrays.asList(expected), actual);
     }
 
+    @Issue("JENKINS-27395")
+    @Test
+    public void enclosingBlocksSingleBlock() throws Exception {
+        rr.addStep(new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                WorkflowJob job = rr.j.createProject(WorkflowJob.class, "enclosingBlocks");
+                job.setDefinition(new CpsFlowDefinition(
+                        "catchError {echo 'bob';}", true));
+
+                WorkflowRun r = rr.j.buildAndAssertSuccess(job);
+
+                FlowExecution execution = r.getExecution();
+
+                // FlowStartNode
+                assertExpectedEnclosing(execution, "2", null);
+
+                // CatchError block, outer block
+                assertExpectedEnclosing(execution, "3", "2");
+
+                // CatchError block, inner block (body)
+                assertExpectedEnclosing(execution, "4", "3");
+
+                // Echo step, enclosed in body block for catchError step
+                assertExpectedEnclosing(execution, "5", "4");
```

Comment:
As mentioned in a long-ago PR, this is going to be pretty fragile; worth spending the time improving the test harness to let you specify step names etc., something like I did with `assertActiveSteps` above. Even with the benefit of a commented-out legend the tests here are close to unreadable.

Reply:
Yeah, I've got some tooling along those lines somewhere - of course now I can't remember where. I'll find it and write a follow-up PR for a harness.

Reply:
Fragile how?  At this point we have so many things tied to the current ID format (including community extensions) that we can't realistically change it.   I actually think the "assertActiveSteps" is quite a bit more fragile because it's reliant on user display strings being immutable and we do have legitimate cases where that might change.

On the other hand, if we want to provide an XPath-like query language for the flowgraph that's something that might have usefulness beyond just testing (though it would require some fairly extensive testcases on its own, which of necessity would have to use something like this).